src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryObjectWriter.cs

```diff
@@ -984,7 +984,7 @@ private long GetAssemblyId(WriteObjectInfo objectInfo)
             {
                 assemId = 0;
             }
-            else if (assemblyString.Equals(Converter.s_urtAssemblyString))
+            else if (assemblyString.Equals(Converter.s_urtAssemblyString) || assemblyString.Equals(Converter.s_urtAlternativeAssemblyString))
```

Comment:
Now I finally understand what you meant. That's definitely something we should mitigate. I think your suggestion, resolving it via the mscorlib shim is correct and is the only option which won't introduce side-effects like you said when we e.g. add new types in S.P.C.

I haven't yet figured out what kind of cache would be appropriate here, for now I will implement it without optimization to test.